src/core/layout/qgslayoutexporter.cpp

```diff
@@ -43,26 +57,693 @@ void QgsLayoutExporter::renderPage( QPainter *painter, int page )
+
+      const QList< QGraphicsItem * > items = mLayout->items();
+      for ( QGraphicsItem *item : items )
+      {
+        mPrevCacheMode.insert( item, item->cacheMode() );
+        item->setCacheMode( QGraphicsItem::NoCache );
+      }
+    }
+
+    ~LayoutItemCacheSettingRestorer()
+    {
+      for ( auto it = mPrevCacheMode.constBegin(); it != mPrevCacheMode.constEnd(); ++it )
+      {
+        it.key()->setCacheMode( it.value() );
+      }
+
+      mLayout->context().mIsPreviewRender = true;
+    }
+
+  private:
+    QgsLayout *mLayout = nullptr;
+    QHash< QGraphicsItem *, QGraphicsItem::CacheMode > mPrevCacheMode;
+};
+
+///@endcond PRIVATE
+
+void QgsLayoutExporter::renderRegion( QPainter *painter, const QRectF &region ) const
 {
   QPaintDevice *paintDevice = painter->device();
   if ( !paintDevice || !mLayout )
   {
     return;
   }
 
-  mLayout->context().mIsPreviewRender = false;
+  LayoutItemCacheSettingRestorer cacheRestorer( mLayout );
+  ( void )cacheRestorer;
 
 #if 0 //TODO
   setSnapLinesVisible( false );
 #endif
 
+  painter->setRenderHint( QPainter::Antialiasing, mLayout->context().flags() & QgsLayoutContext::FlagAntialiasing );
+
   mLayout->render( painter, QRectF( 0, 0, paintDevice->width(), paintDevice->height() ), region );
 
 #if 0 // TODO
   setSnapLinesVisible( true );
 #endif
+}
+
+QImage QgsLayoutExporter::renderRegionToImage( const QRectF &region, QSize imageSize, double dpi ) const
+{
+  double resolution = mLayout->context().dpi();
+  double oneInchInLayoutUnits = mLayout->convertToLayoutUnits( QgsLayoutMeasurement( 1, QgsUnitTypes::LayoutInches ) );
+  if ( imageSize.isValid() )
+  {
+    //output size in pixels specified, calculate resolution using average of
+    //derived x/y dpi
+    resolution = ( imageSize.width() / region.width()
+                   + imageSize.height() / region.height() ) / 2.0 * oneInchInLayoutUnits;
+  }
+  else if ( dpi > 0 )
+  {
+    //dpi overridden by function parameters
+    resolution = dpi;
+  }
+
+  int width = imageSize.isValid() ? imageSize.width()
+              : static_cast< int >( resolution * region.width() / oneInchInLayoutUnits );
+  int height = imageSize.isValid() ? imageSize.height()
+               : static_cast< int >( resolution * region.height() / oneInchInLayoutUnits );
+
+  QImage image( QSize( width, height ), QImage::Format_ARGB32 );
+  if ( !image.isNull() )
+  {
+    image.setDotsPerMeterX( resolution / 25.4 * 1000 );
+    image.setDotsPerMeterY( resolution / 25.4 * 1000 );
+    image.fill( Qt::transparent );
+    QPainter imagePainter( &image );
+    renderRegion( &imagePainter, region );
+    if ( !imagePainter.isActive() )
+      return QImage();
+  }
+
+  return image;
+}
+
+///@cond PRIVATE
+class LayoutContextSettingsRestorer
+{
+  public:
+
+    LayoutContextSettingsRestorer( QgsLayout *layout )
+      : mLayout( layout )
+      , mPreviousDpi( layout->context().dpi() )
+      , mPreviousFlags( layout->context().flags() )
```

Comment:
We could check that `layout` is not null before using it.

Reply:
Again, there's no use for this class if layout is null. If it is, an assert would be better... but we would have crashed long before this class is used anyway :)

Reply:
Fixed